[BEAM-13990] BigQueryIO cannot write to DATE and TIMESTAMP columns when using Storage Write API

liu-du opened the pull request:
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

liu-du commented:
R: @reuvenlax 

Hi Reuven Lax, I tried using BigQueryIO.withMethod(BigQueryIO.Write.Method.STORAGE_WRITE_API) and it turned out DATE and TIMESTAMP values were converted to the correct type. I've change the type conversion according to [write api documentation](https://cloud.google.com/bigquery/docs/write-api#data_type_conversions) (DATE values need to be converted to protobuf int32 and TIMESTAMP values to int64).

reuvenlax commented:
I don't think this quite works - you would need code to convert the JSON value (which is probably a string) to the numeric value. In any case, BigQuery should be accepting string values here - formatting as an integer should be a performance change only.

liu-du commented:
> BigQuery should be accepting string values here - formatting as an integer should be a performance change only

I think it's a _correctness_ issue rather than performance change. BigQuery Storage Write API _rejects_ date/timestamp encoded in string format. In this commit [b56823d1d213adf6ca5564ce1d244cc4ae8f0816](https://github.com/liu-du/beam/commit/b56823d1d213adf6ca5564ce1d244cc4ae8f0816), I used existing conversion (use string value for date/timestamp columns) and add an integration test to write date/timestamp using protobuf string values, the error I get is:

`INVALID_ARGUMENT: The proto field mismatched with BigQuery field at Da40f0078_2574_4708_b66f_79eb0e516c43.datevalue, the proto field type string, BigQuery field type DATE Entity`

reuvenlax commented:
The BigQuery team has stated that this should work (writing as strings), so this may be a bug in BigQuery.

Regardless, I don't see any code in this PR to convert the strings to the numeric values. Did a commit get dropped?

liu-du commented:
No I didn't add code to convert strings to int/long. With current code change, user should pass in int/long when constructing TableRow, like is done in the test [here](https://github.com/apache/beam/pull/16926/files#diff-91c88a463f78112ec572fc5230d9220882b0d2179fc15b057b8be17ff4b9c820R97-R100).

liu-du commented:
Hmm, the problem is in `TableRowToStorageApiProto.scalarToProtoValue`. Let's say it needs to convert "2019-08-16", all `scalarToProtoValue` knows is that it is a protobuf string, it doesn't know whether it's actually a BigQuery DATE or it's just a BigQuery STRING... To make it work, we need to also pass in BigQuery type to `scalarToProtoValue`, so it's signature needs to change from `scalarToProtoValue(FieldDescriptor fieldDescriptor, Object jsonBQValue)` to `scalarToProtoValue(TableFieldSchema tableFieldSchema, FieldDescriptor fieldDescriptor, Object jsonBQValue)`.

What do you think of this approach?

reuvenlax commented:
Yeah, you can pass it down, though you'll have to do so recursively I think

reuvenlax commented:
Let me know when this is ready for review

liu-du commented:
@reuvenlax 

This is ready for review. A brief summary of changes:

- TableFieldSchema is now passed down recursively to correctly covert date/timestamp types. 
- Added JavaTimeModule in [TableRowJsonCoder.java](https://github.com/apache/beam/pull/16926/files#diff-0bc4c840814c5cf4da045f15a873defedb555828faf9986a0fbd3a66c3e81279) to enable jackson to handle date/time etc.
- Allow more types to be passed into TableRow (e.g. pass a java.time.Instant to a timestamp field). Type conversion is handled in TableRowToStorageApiProto.scalarToProtoValue and tested in TableRowToStorageApiProtoIT.testTableRowToStorageApiProtoITRichType
- Removed the destinationDescriptorCache in StorageApiDynamicDestinationsTableRow.java

reuvenlax commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinationsTableRow.java:
```diff
@@ -23,24 +23,18 @@
 
-  // TODO: Is this cache needed? All callers of getMessageConverter are already caching the resullt.
-  private final Cache<DestinationT, Descriptor> destinationDescriptorCache =
-      CacheBuilder.newBuilder().expireAfterAccess(Duration.ofMinutes(15)).build();
```
Can you explain why this cache was removed? Is it related to this change?

liu-du replied:
I agree with the TODO comment that the cache is not needed here. The MessageConverter return by getMessageConverter also need to remember the equivalent TableSchema, but this cache only saves Descriptor, so I removed it entirely

reuvenlax commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoIT.java:
```diff
@@ -0,0 +1,283 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.bigquery;
+
```
This is great! Make sure to run this test on the the Dataflow runner as well.

reuvenlax commented:
FYI - I spoke to the BigQuery team, and they believe that it's a bug that they don't support strings for DATE fields and will be fixing it. That being said, this PR is still worthwhile as it makes our proto encoding smaller.

liu-du commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:
```diff
@@ -196,102 +222,131 @@ private static Object messageValueFromFieldValue(
+        } else if (jsonBQValue instanceof Double) {
+          return jsonBQValue;
+        } else if (jsonBQValue instanceof Float) {
+          return ((Float) jsonBQValue).longValue();
```
good catch, it's a typo...

liu-du commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowJsonCoder.java:
```diff
@@ -68,7 +73,17 @@ public long getEncodedElementByteSize(TableRow value) throws Exception {
+          .configure(JsonGenerator.Feature.WRITE_BIGDECIMAL_AS_PLAIN, true)
+          .withConfigOverride(
+              BigDecimal.class,
+              it -> it.setFormat(JsonFormat.Value.forShape(JsonFormat.Shape.STRING)))
```
I changed it because Jackson serializes BigDecimal as float by default, but floats are prone to lose precision. this can be avoided by serializing BigDecimal to string.

The new configuration only changes how Jackson serialize TableRows, deserialization should continue to work since it's just deserializing plain json. Still, the in-flight BigDecimal (already serialized as floats) may need to be converted to protobuf String, for this I've added some code in `TableRowToStorageApiProto.scalarToProtoValue` to convert double/floats to String. I hope this should be enough to address backward compatibility?

liu-du commented:
@reuvenlax Thanks for the review. 

I've made changes to address the issues flagged, please have a look at the code changes. I still think the cache is not needed:
1. `TwoLevelMessageConverterCache` already cache MessageConverter twice.
2. code will be very different even if we add back caching, because it needs to cache TableSchema in addition to Descriptor.

aaltay commented:
@liu-du - Could you please address the failing tests?

liu-du commented:
@aaltay most tests is passing now except "CommunityMetrics", the error message is something like `Unable to connect to the server: dial tcp 104.154.102.21:443: i/o timeout`. I'm not sure what's going on, could you or anyone help?

kileys commented:
Community metrics failures is a known issue and unrelated to your changes

reuvenlax commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:
```diff
@@ -182,116 +180,164 @@ private static void fieldDescriptorFromTableField(
-      case BOOL:
-        if (jsonBQValue instanceof Boolean) {
-          return jsonBQValue;
+  static Object scalarToProtoValue(
```
I'm not sure  I consider a message type  a scalar

reuvenlax commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:
```diff
@@ -182,116 +180,164 @@ private static void fieldDescriptorFromTableField(
+          // joda instant precision is millisecond
+          return ((org.joda.time.Instant) value).getMillis() * 1000L;
+        } else if (value instanceof Integer || value instanceof Long) {
+          return ((Number) value).longValue();
```
why here you assume the number is micros but below you assume it's seconds?

reuvenlax commented:
@liu-du Friendly ping. If you don't have time, I can try and resolve the conflicts and merge this PR

aaltay commented:
> @liu-du Friendly ping. If you don't have time, I can try and resolve the conflicts and merge this PR

Thank you @reuvenlax. I think that would be helpful.

reuvenlax commented:
@aaltay I resolved conflicts into https://github.com/apache/beam/pull/17404

aaltay commented:
(Closing this. #17404 is the updated version.)
